Gitweb:     http://git.kernel.org/git/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commit;h=fe6c700ff34e68e1eb7991e9c5d18986d0005ac1
Commit:     fe6c700ff34e68e1eb7991e9c5d18986d0005ac1
Parent:     bda1cda54b7e1654e9416fec76e6342847b25216
Author:     Ralph Loader <suckfish@ihug.co.nz>
AuthorDate: Mon Sep 22 21:06:48 2008 -0300
Committer:  Mauro Carvalho Chehab <mchehab@redhat.com>
CommitDate: Sat Oct 4 23:04:32 2008 -0300

    V4L/DVB (9053): fix buffer overflow in uvc-video
    
    There is a buffer overflow in drivers/media/video/uvc/uvc_ctrl.c:
    
    INFO: 0xf2c5ce08-0xf2c5ce0b. First byte 0xa1 instead of 0xcc
    INFO: Allocated in uvc_query_v4l2_ctrl+0x3c/0x239 [uvcvideo] age=13 cpu=1 pid=4975
    ...
    
    A fixed size 8-byte buffer is allocated, and a variable size field is read
    into it; there is no particular bound on the size of the field (it is
    dependent on hardware and configuration) and it can overflow [also
    verified by inserting printk's.]
    
    The patch attempts to size the buffer to the correctly.
    
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Acked-by: Laurent Pinchart <laurent.pinchart@skynet.be>
    Signed-off-by: Mauro Carvalho Chehab <mchehab@redhat.com>
---
 drivers/media/video/uvc/uvc_ctrl.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/media/video/uvc/uvc_ctrl.c b/drivers/media/video/uvc/uvc_ctrl.c
index 6ef3e52..feab12a 100644
--- a/drivers/media/video/uvc/uvc_ctrl.c
+++ b/drivers/media/video/uvc/uvc_ctrl.c
@@ -592,7 +592,7 @@ int uvc_query_v4l2_ctrl(struct uvc_video_device *video,
 	if (ctrl == NULL)
 		return -EINVAL;
 
-	data = kmalloc(8, GFP_KERNEL);
+	data = kmalloc(ctrl->info->size, GFP_KERNEL);
 	if (data == NULL)
 		return -ENOMEM;
 

